### PR TITLE
Count only visible repos on profile

### LINKED ISSUE #25914: Number of repositories includes the number of private repositories in profile page
### Description

The number of repositories displayed in the `Repositories` tab contains the number of private repositories.

### Gitea Version

1.20.0

### Can you reproduce the bug on the Gitea demo site?

Not tested (I would have if there were no captcha)

### Log Gist

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Self built 1.20.0 with bindata (not patched)

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by JakobDev**
Fixes #25914

This should probably be backported

### CODE REVIEW

**Comment by JakobDev** [comment]
> Instead of adding another template property, couldn't we just show the length of the to-be-rendered repos?

This will not work. The repo count is always shown, no matter in which tab you are. But the list of repos is only loaded, when you are in the repos tab.

**Comment by silverwind** [comment]
Right, so only on the repo tab we could skip the count, hmm.

**Comment by silverwind** [comment]
How long does the extra code it take with a moderate amount of repos? That should give us an estimation whether it is worth it.

**Review by silverwind — APPROVED** [review]

**Comment by JakobDev** [comment]
I found out that the Bar rendered by a few more routes, so I decided to move the Code into a new function.

How long does the extra code it take with a moderate amount of repos?

What exactly did you mean with "moderate amount of repos"? 50? 100? 500? Counting in SQL is relay fast, so I don't think it matters that much, even with more Repos.

I also realized, that the same Problem apers in the Stars Tab. If you star an private Repo, you have for Visitors a higher Star Number than actually shown in the List.

**Comment by JakobDev** [comment]
I know realized this is harder than I though: If you give someone access to a private Repo, the private Repo will be shown when the other Person visits your profile, so the Count would be wrong. I have to find an alternative.

**Comment by silverwind** [comment]
> What exactly did you mean with "moderate amount of repos"? 50? 100? 500? Counting in SQL is relay fast, so I don't think it matters that much, even with more Repos.

Yeah, likely not an issue. I recall at least in MySQL `COUNT()` is `O(n)` but given that repo numbers are at most in the thousands, it won't be an issue.

**Comment by silverwind** [comment]
Indeed, this will need a complex query.

**Comment by JakobDev** [comment]
Can we run this on every profile visit or does this take to long just for displaying a number?

**Review by puni9869 — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Seems good to merge. Not sure we should backport it though.

**Comment by JakobDev** [comment]
This is a fix for a Bug hat was reported for 1.20 and doesn't introduce new Features, so I would say that it should be backported

File: routers/web/shared/user/header.go [hunk 1]
```diff
@@ -110,3 +111,21 @@ func RenderUserHeader(ctx *context.Context) {
 	defer profileClose()
 	ctx.Data["HasProfileReadme"] = profileReadmeBlob != nil
 }
+
+func LoadHeaderCount(ctx *context.Context) error {
+	prepareContextForCommonProfile(ctx)
+
+	_, repoCount, err := repo_model.SearchRepository(ctx, &repo_model.SearchRepoOptions{
```
Comment (lunny) [inline routers/web/shared/user/header.go]: No, you need a count function, but not use search function here.

**Review by lunny — DISMISSED** [review]
Don't use search function as count function, performance problem.

File: routers/web/shared/user/header.go (hunk 1 above)
Comment (JakobDev) [reply routers/web/shared/user/header.go]: Is there a Count function which checks if a User is able to see a Repo?

Comment (lunny) [reply routers/web/shared/user/header.go]: Looks like there is not, but you can write one like this.
```go
func CountRepository(opts *repo_model.SearchRepoOptions) (int64,error) {
     return sess.
			Where(repo_model.SearchRepositoryCondition(opts)).
			Count(new(Repository))
}
```

Comment (JakobDev) [reply routers/web/shared/user/header.go]: Done

**Review by lunny — APPROVED** [review]

**Comment by JakobDev** [comment]
```
$ go run ./contrib/backport 25928
Unable to read `docs/config.yaml`: open docs/config.yaml: no such file or directory
* Backporting 25928 to origin/release/v1.18 as backport-25928-v1.18
* `git fetch origin main`

* `git fetch origin release/v1.18`

Unable to backport: unable to fetch release/v1.18 from origin: exit status 128
$ 
```

**Pull request merged**